src/libwhb/src/commandserver.c

```diff
@@ -0,0 +1,123 @@
+#include <coreinit/baseheap.h>
+#include <coreinit/expandedheap.h>
+#include <nsysnet/socket.h>
+#include <whb/commandserver.h>
+#include <whb/libmanager.h>
+#include <whb/log.h>
+
+#include <string.h>
+
+#define SERVER_PORT 4406
+
+int
+sSocket = -1;
+
+struct sockaddr_in
+sAddr;
+
+void
+closeSocket(const char * funcName)
+{
+   int ret = socketclose(sSocket);
+   if(ret < 0) {
+      WHBLogPrintf("%s: Error occurred closing socket: %d", funcName, socketlasterr());
+   }
+}
+
+void
+closeClient(const char * funcName, int fd)
+{
+   int ret = socketclose(fd);
+   if(ret < 0) {
+      WHBLogPrintf("%s: Error occurred closing client socket: %d", funcName, socketlasterr());
+   }
+}
+
+BOOL
+WHBCommandServerInit()
+{
+   int ret = 0;
+
+   if(sSocket >= 0) {
+      WHBLogPrintf("%s: Command server is already running.", __FUNCTION__);
+      return TRUE;
+   }
+
+   WHBInitializeSocketLibrary();
+
+   sSocket = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
+   if(sSocket < 0) {
+      WHBLogPrintf("%s: Error occurred while creating socket: %d", __FUNCTION__, socketlasterr());
+      sSocket = -1;
+      return FALSE;
+   }
+
+   memset(&sAddr, 0, sizeof(struct sockaddr_in));
+   sAddr.sin_family = AF_INET;
+   sAddr.sin_port = htons(SERVER_PORT);
+   sAddr.sin_addr.s_addr = htonl(INADDR_ANY);
+
+   ret = bind(sSocket, (struct sockaddr *)&sAddr, sizeof(struct sockaddr_in));
+   if(ret < 0) {
+      WHBLogPrintf("%s: Error occurred while binding to socket: %d", __FUNCTION__, socketlasterr());
+      closeSocket(__FUNCTION__);
+      sSocket = -1;
+      return FALSE;
+   }
+
+   ret = listen(sSocket, 3);
+   if(ret < 0) {
+      WHBLogPrintf("%s: Error occurred while setting socket to listen mode: %d", __FUNCTION__, socketlasterr());
+      closeSocket(__FUNCTION__);
+      sSocket = -1;
+      return FALSE;
+   }
+
+   return TRUE;
+}
+
+void
+WHBCommandServerStop()
+{
+   if(sSocket < 0) {
+      WHBLogPrintf("%s: Socket is already closed.", __FUNCTION__);
+      return;
+   }
+
+   closeSocket(__FUNCTION__);
+   sSocket = -1;
+   WHBDeinitializeSocketLibrary();
+}
+
+BOOL
+WHBCommandServerListen(char * stringLocation)
```

Comment:
For a command server would it be better to keep the connection open rather than listen, accept, read, close, for every single request? I'm not opposed to accepting this as is, but if I personally did it I would write it as a persistent connection.

Reply:
Yeah, that definitely is a better idea. I originally planned to write this using UDP, in which this kind of connection model is not necessary, but it didn't really work.